### DISCUSSION TITLE
Trained layers with LoRA will be merged into model?

### DISCUSSION THREAD

**Discussion by lavinal712** [huggingface/peft#1727] [Q&A]
If I train a LoRA, and want to save some layers in safetenser, like this:
```
config = LoraConfig(
            r=args.lora_r,
            lora_alpha=args.lora_alpha,
            target_modules=TARGET_MODULE,
            lora_dropout=args.lora_dropout,
            bias=args.lora_bias,
            modules_to_save=SAVED_MODULE,
        )
       model  = get_peft_model(model, config)
```
Then I conduct:
```
model = PeftModel.from_pretrained(model, "model")
model = model.merge_and_unload()
model.save_pretrained(save_path)
```
Trained layers with LoRA will be merged into model? Or just thrown away?

**Comment by BenjaminBossan** [comment]
`model.merge_and_unload()` will merge the LoRA weights into the base weights. If you want to throw them away, use `model.unload()`.

**Comment by lavinal712** [comment]
Well, I mean I trained some layers (not lora) with lora, and when I use merge_and_unload(), will the layers be updated?

**Reply by BenjaminBossan** [reply]
No, only the LoRA weights should be updated. Therefore, when merging, only the layers that have LoRA applied should change. All other layers should be the same.

**Reply by lavinal712** [reply]
thank, I will try to adjust my code to fix it.

**Reply by BenjaminBossan** [reply]
If something doesn't seem to be working as expected, feel free to paste the code so that I can take a look.

**Comment by lavinal712** [comment]
I write this to merge lora and trainable layers to model, but failed.
```
CONTROLNET_TARGET_MODULE = ["to_q", "to_k", "to_v", "to_out.0",
                            "conv", "conv_in", "conv_out", "conv_shortcut", "conv1", "conv2",
                            "proj", "proj_in", "proj_out", "time_emb_proj",
                            "linear_1", "linear_2", "ff.net.2"]
CONTROLNET_SAVED_MODULE = ["controlnet_cond_embedding.blocks.0", "controlnet_cond_embedding.blocks.1",
                           "controlnet_cond_embedding.blocks.2", "controlnet_cond_embedding.blocks.3",
                           "controlnet_cond_embedding.blocks.4", "controlnet_cond_embedding.blocks.5",
                           "controlnet_down_blocks.0", "controlnet_down_blocks.1", "controlnet_down_blocks.2",
                           "controlnet_down_blocks.3", "controlnet_down_blocks.4", "controlnet_down_blocks.5",
                           "controlnet_down_blocks.6", "controlnet_down_blocks.7", "controlnet_down_blocks.8",
                           "controlnet_down_blocks.9", "controlnet_down_blocks.10", "controlnet_down_blocks.11",
                           "controlnet_mid_block"]
...
    if args.use_lora:
        config = LoraConfig(
            r=args.lora_r,
            lora_alpha=args.lora_alpha,
            target_modules=CONTROLNET_TARGET_MODULE,
            lora_dropout=args.lora_dropout,
            bias=args.lora_bias,
            modules_to_save=CONTROLNET_SAVED_MODULE,
        )
        controlnet = get_peft_model(controlnet, config)
        controlnet.print_trainable_parameters()
...
```
```
from collections import OrderedDict
import torch
import torch.nn as nn

import diffusers
from diffusers import UNet2DConditionModel, ControlNetModel
from peft import PeftModel
from safetensors.torch import load_file


def _tie_weights(source_state_dict, target_state_dict):
    for k in source_state_dict.keys():
        if k.startswith("base_model.model."):
            new_k = k[len("base_model.model."):]
            if new_k in target_state_dict.keys():
                print(new_k)
                target_state_dict[new_k] = source_state_dict[k]
    return target_state_dict


base_model = "F:/stable-diffusion-v1-5"
controlnet_save_path = "./save"

unet = UNet2DConditionModel.from_pretrained(
    base_model, subfolder="unet", torch_dtype=torch.float16
)

controlnet = ControlNetModel.from_unet(unet)
controlnet_state_dict = controlnet.state_dict()
lora_state_dict = load_file("controlnet/adapter_model.safetensors")
controlnet_state_dict = _tie_weights(lora_state_dict, controlnet_state_dict)
controlnet.load_state_dict(controlnet_state_dict)
controlnet = PeftModel.from_pretrained(controlnet, "controlnet")
controlnet = controlnet.merge_and_unload()
controlnet.save_pretrained(controlnet_save_path)
```
The output model does not have control ability. I have trained it for 60000 steps with batch size of 4.
I wonder how to fix it.

**Reply by BenjaminBossan** [reply]
A few points:

1. Does the net work correctly if you don't call `merge_and_unload`?
2. You have quite a lot of entries for `modules_to_save`, is there a specific reason for that?
3. You have tied weights. Merging into tied weights can be problematic. For example, if multiple different LoRA weights are merged into the same weight because it is tied, you will probably get incorrect results.

**Reply by lavinal712** [reply]
1. I didn't do that. But I tested the merged model with sd-webui, the result was like using sd model and there was no control in image if  I send the conditional_images. I believe it works like the air does.
2. I refer a repo [control-lora-v2](https://github.com/HighCWu/control-lora-v2), the model ControlLoRA's training layers are like that. I have trained a ControlLoRA model and it works well (control) .Maybe it is not so necessary but for now there is no big deals as ControlLoRA works but does not mine.
3. I found tied weights is not necessary. merge_and_unload() can bind trainable layers to the model and merge lora.

The main purpose of my code is to rewrite the repo [control-lora-v2](https://github.com/HighCWu/control-lora-v2) by peft, and adding some new finetuning methods.

**Reply by BenjaminBossan** [reply]
> 1. I didn't do that.

If you could give it a try, that would be helpful. That way we know if the merging is the problem or something else.

> I believe it works like the air does.

I don't understand that.

> I found tied weights is not necessary.

Could you try what happens when you merge without tying the weights first?

**Reply by lavinal712** [reply]
I tried tie weights wihtout merge, there is result:
![image](https://github.com/huggingface/peft/assets/98888959/bc7b907b-5892-4649-8496-a3fa25170d58)

**Reply by lavinal712** [reply]
Here you can see there is no condition in image. I trained the model for 40000  steps with batch size of 4.

**Reply by lavinal712** [reply]
here is after merge_and_unload() w/o tie
![image](https://github.com/huggingface/peft/assets/98888959/bbdac21e-0c54-4a02-a10e-baf0dfb6ece8)
still no control

**Reply by BenjaminBossan** [reply]
I don't know much about controlnet, so not sure what I should be looking for. I assume you have some reference image but the generated images don't resemble that reference image.

Anyway, if the model does not deliver expected results even _without_ merging, it means that merging is not the issue. Instead, it appears like the model is not trained correctly. Is that observation correct?

**Reply by lavinal712** [reply]
Thanks for your help! I find my lr wae set 1e-5, which suggests that the model may converge slowly. After I change the lr, the model come to have control ability.
